### PR TITLE
Added support for Selectors

### LINKED ISSUE #7: Composing Selector: How @Select work with composing selector?
I have some selectors that composed from URL and entities, how can I convert it to use with `@Select`?

Can we do something like this?

```ts
export function Select(selector?: string | MemoizedSelector<any, any>): any {
  return function(target: any, name: string, descriptor: TypedPropertyDescriptor<any>): void {
    if (delete target[name] && typeof selector === 'string') {
      Object.defineProperty(target, name, {
        get: () => {
          if (!NgrxSelect.store) {
            throw new Error('NgrxSelect not connected to store!');
          }

          const fn = memoize(state => getValue(state, selector || name));
          return NgrxSelect.store.select(fn);
        },
        enumerable: true,
        configurable: true,
      });
    } else if (delete target[name] && typeof selector === 'function') {
      Object.defineProperty(target, name, {
        get: () => {
          if (!NgrxSelect.store) {
            throw new Error('NgrxSelect not connected to store!');
          }
          return NgrxSelect.store.select(selector);
        },
        enumerable: true,
        configurable: true,
      });
    }
  };
}

```

### LINKED ISSUE #15: ActionType should be generic
current implement: 
`symbols.ts`
```ts
export type ActionType = { new (...args: any[]): Action };
```

We can add generic type for ActionType like this:

```ts
export type ActionType<T extends Action = Action> = { new (...args: any[]): T };
```

Then we can reuse it, i.e in of-action.ts

```ts
import { ActionType } from './symbols';

export function ofAction<T extends Action>(allowedType: ActionType<T>): OperatorFunction<Action, T>;
```

or in action.ts

```ts
export function Action(...actionsKlasses: ActionType[]) { ... }
```

### LINKED ISSUE #16: Compatible with strict null check?
Since Angular 5 is compatible with strictNullCheck, can you consider adding strict: true to tsconfig.json?

### LINKED ISSUE #17: Better infer type checking ofAction
Can you consider changing return type OperatorFunction<Action, Action> to OperatorFunction<Action, T>?

https://github.com/amcdnl/ngrx-actions/blob/6ed55d5d9d477b9557afe42021c68ab502a39f66/src/of-action.ts#L10

Then user could use like this

```ts
@Effect()
getFooBar$ = this.actions$.pipe(
    ofAction<GetFoo | GetBar>(
      GetFoo, GetBar
    ),
    map(action /* action has infer type GetFoo | GetBar */ => { ... })
```

### PR DESCRIPTION
**Pull request by albyrock87**
Unit Tested `@Select`
Added support for Selectors
Speedup generated selector from path with deeply nested props
Added support for ('this', 'is', 'a', 'splitted', 'path')
Closes #7 
Closes #15 
Closes #17 
Closes #16

### CODE REVIEW

File: src/index.spec.ts [hunk 1]
```diff
@@ -131,13 +137,72 @@ describe('actions', () => {
...
+    class MyAction3 implements NgRxAction {
+      readonly type = 'myaction3';
+      constructor(public foo: any, public bar: any) {}
+    }
+
+    const action = new MyAction('foo'),
+      action2 = new MyAction2(),
+      action3 = new MyAction3('a', 0);
+    const actions = of<NgRxAction>(action, action2, action3);
+    let tappedActions: NgRxAction[] = [];
```
Comment (sandangel) [inline src/index.spec.ts]: use const instead, lint error here

File: src/index.spec.ts [hunk 2]
```diff
@@ -131,13 +137,72 @@ describe('actions', () => {
...
+                d: 'world'
+              }
+            }
+          }
+        }
+      }
+    };
+
+    const msFeature = createFeatureSelector<FooState>('myFeature');
+    const msBar = createSelector(msFeature, state => state.bar);
```
Comment (sandangel) [inline src/index.spec.ts]: lint error: `Shadowed name`.

File: src/select.ts [hunk 3]
```diff
@@ -11,17 +10,42 @@ export class NgrxSelect {
...
+    }
+    // Handle string vs Selector<TState, TValue>
+    if (typeof selectorOrFeature === 'string') {
+      if (paths.length) {
+        selectorOrFeature = [selectorOrFeature, ...paths].join('.');
+      }
+      fn = fastPropGetter(selectorOrFeature);
+    } else {
+      fn = selectorOrFeature;
+    }
```
Comment (sandangel) [inline src/select.ts]: why do we have to `.join('.')`, then `.split('.')`. 
how about 
```ts
function fastPropGetter(paths: string[]): (x: any) => any {
  const segments = paths;
  let seg = 'store.' + segments[0],
    i = 0;
  const = segments.length;
  let expr = seg;
  while (++i < l) {
    expr = expr + ' && ' + (seg = seg + '.' + segments[i]);
  }
  const fn = new Function('store', 'return ' + expr + '');
  return <(x: any) => any>fn;
}
```
then 
```ts
if (typeof selectorOrFeature === 'string' && selectorOrFeature /* should throw error with empty string */) {
      let propsArray = [selectorOrFeature, ...paths];
      if (!paths.length) {
        propsArray = selectorOrFeature.split('.');
      }
      fn = fastPropGetter(propsArray);
    } else if (typeof selectorOrFeature === 'function') {
      fn = selectorOrFeature;
    } else {
       throw new Error('...')
    }
```

Comment (sandangel) [reply src/select.ts]: we should check case when user put a dot in selectorOrFeature and paths.length > 0 too

File: src/internals.ts [hunk 4]
```diff
@@ -1,7 +1,8 @@
 import { Action } from '@ngrx/store';
```
Comment (sandangel) [inline src/internals.ts]: This import is no longer needed

File: src/select.ts [hunk 5]
```diff
@@ -11,17 +10,42 @@ export class NgrxSelect {
...
-  return function(target: any, name: string, descriptor: TypedPropertyDescriptor<any>): void {
+export function Select<TState = any, TValue = any>(
+  selector: Selector<TState, TValue>
+): (target: any, name: string) => void;
+export function Select<TState = any, TValue = any>(
+  selectorOrFeature?: string,
+  ...paths: string[]
+): (target: any, name: string) => void;
+export function Select<TState = any, TValue = any>(
+  selectorOrFeature: string | Selector<TState, TValue>,
```
Comment (sandangel) [inline src/select.ts]: overload signature is not compatible with function implementation,

should be `selectorOrFeature?: string | ...`

File: src/index.spec.ts [hunk 6]
```diff
@@ -131,13 +137,72 @@ describe('actions', () => {
...
+
+    class MyStateSelector {
+      @Select('myFeature.bar.a.b.c.d') hello$: Observable<string>; // deeply nested props
+      @Select() myFeature: Observable<FooState>; // implied by name
+      @Select(msBar) bar$: Observable<any>; // using MemoizedSelector
+    }
+
+    const store = new NgRxStore(of(state), undefined, undefined);
```
Comment (sandangel) [inline src/index.spec.ts]: not compatible with strictNullCheck. Consider using tsconfig.spec.json file for testing only

File: src/select.ts [hunk 7]
```diff
@@ -30,9 +54,22 @@ export function Select(path?: string): any {
...
+  const segments = path.split('.');
+  let seg = 'store.' + segments[0],
+    i = 0,
+    l = segments.length;
+  let expr = seg;
+  while (++i < l) {
+    expr = expr + ' && ' + (seg = seg + '.' + segments[i]);
   }
-  return state;
+  const fn = new Function('store', 'return ' + expr + '');
```
[the remark this answers is not in the document]
Comment (albyrock87) [reply src/select.ts]: I forgot to remove the empty string here. ';' is not needed.

**Comment by sandangel** [comment]
this PR will closes #16 too.

**Comment by amcdnl** [comment]
Beautiful!

**Comment by amcdnl** [comment]
Release 2.1.0!

**Pull request merged**